File: bench/src/loom_otp_native/bench/spawn.clj
(ns loom-otp.bench.spawn
  "Spawn benchmarks for loom-otp."
  (:require [loom-otp.process :as proc]
            [loom-otp.process.match :refer [receive!]]
            [loom-otp.bench.common :as c]
            [mount.lite :as mount]))

;; =============================================================================
;; Spawn Benchmarks
;; =============================================================================

(defn spawn-n-immediate-exit
  "Spawn N processes that exit immediately (return :done).
   Measures raw spawn throughput without synchronization."
  [n]
  (let [time-ms (c/time-ms
                  (dotimes [_ n]
                    (proc/spawn! :done)))]
    (c/println-result (format "spawn-immediate-exit n=%d" n)
                      (/ n (/ time-ms 1000.0))
                      "proc/sec")))

(defn spawn-n-async
  "Spawn N processes asynchronously, fire-and-forget.
   Measures pure async spawn throughput without waiting for completion."
  [n]
  (let [time-ms (c/time-ms
                  (->> (range n)
                       (mapv (fn [_] (proc/spawn-async! :done)))))]
    (c/println-result (format "spawn-async n=%d" n)
                      (/ n (/ time-ms 1000.0))
                      "proc/sec")))

(defn spawn-n-wait-exit
  "Spawn N processes, wait for all to exit using counter.
   Measures spawn + exit coordination throughput."
  [n]
  (let [done (promise)
        counter (atom 0)
        time-ms (c/time-ms
                  (dotimes [_ n]
                    (proc/spawn!
                      (when (= n (swap! counter inc))
                        (deliver done true))))
                  (deref done 60000 :timeout))]
    (c/println-result (format "spawn-wait-exit n=%d" n)
                      (/ n (/ time-ms 1000.0))
                      "proc/sec")))

(defn spawn-sequential
  "Spawn one process, wait for it to exit, repeat N times.
   Measures spawn latency (round-trip time)."
  [n]
  (let [time-ms (c/time-ms
                  (dotimes [_ n]
                    (let [done (promise)]
                      (proc/spawn! (deliver done true))
                      (deref done 1000 :timeout))))]
    (c/println-result (format "spawn-sequential n=%d" n)
                      (/ time-ms n)
                      "ms/spawn")))

(defn spawn-tree
  "Create binary tree of processes, depth D -> 2^(D+1)-1 processes.
   Each node spawns two children, waits for results, reports sum to parent."
  [depth]
  (letfn [(node [parent d]
            (proc/spawn!
              (if (zero? d)
                (proc/send parent 1)
                (let [self (proc/self)]
                  (node self (dec d))
                  (node self (dec d))
                  (receive! n1
                    (receive! n2
                      (proc/send parent (+ 1 n1 n2))))))))]
    (let [result (promise)
          time-ms (c/time-ms
                    (proc/spawn!
                      (node (proc/self) depth)
                      (receive! total (deliver result total)))
                    (deref result 120000 :timeout))
          total @result]
      (c/println-result (format "spawn-tree depth=%d procs=%d" depth total)
                        (/ total (/ time-ms 1000.0))
                        "proc/sec"))))

(defn spawn-link-tree
  "Same as spawn-tree but with linked processes.
   Measures link setup overhead."
  [depth]
  (letfn [(node [parent d]
            (proc/spawn-link!
              (if (zero? d)
                (proc/send parent 1)
                (let [self (proc/self)]
                  (node self (dec d))
                  (node self (dec d))
                  (receive! n1
                    (receive! n2
                      (proc/send parent (+ 1 n1 n2))))))))]
    (let [result (promise)
          time-ms (c/time-ms
                    (proc/spawn!
                      (node (proc/self) depth)
                      (receive! total (deliver result total)))
                    (deref result 120000 :timeout))
          total @result]
      (c/println-result (format "spawn-link-tree depth=%d procs=%d" depth total)
                        (/ total (/ time-ms 1000.0))
                        "proc/sec"))))

;; =============================================================================
;; Runner
;; =============================================================================

(defn- warm-up-spawn! []
  (doseq [n c/SPAWN-SCALES]
    (spawn-n-immediate-exit n))
  (doseq [n c/SPAWN-SCALES]
    (spawn-n-wait-exit n))
  (doseq [d c/TREE-DEPTHS]
    (spawn-tree d)))

(defn run-all []
  (mount/start)
  (try
    (c/println-header "SPAWN BENCHMARKS")
    
    ;; Warm-up phase
    (c/warm-up! warm-up-spawn!)
    
    (c/println-subheader "Immediate Exit (fire-and-forget)")
    (doseq [n c/SPAWN-SCALES]
      (spawn-n-immediate-exit n))
    
    (c/println-subheader "Async Spawn")
    (doseq [n c/SPAWN-SCALES]
      (spawn-n-async n))
    
    (c/println-subheader "Wait for Exit (synchronized)")
    (doseq [n c/SPAWN-SCALES]
      (spawn-n-wait-exit n))
    
    (c/println-subheader "Sequential (latency)")
    (doseq [n c/SPAWN-SCALES-SMALL]
      (spawn-sequential n))
    
    (c/println-subheader "Process Tree")
    (doseq [d c/TREE-DEPTHS]
      (spawn-tree d))
    
    (c/println-subheader "Linked Process Tree")
    (doseq [d c/TREE-DEPTHS]
      (spawn-link-tree d))
    
    (finally
      (mount/stop))))

File: bench/src/elixir/lib/bench_elixir/bench/spawn.ex
defmodule BenchElixir.Bench.Spawn do
  @moduledoc """
  Spawn benchmarks for Elixir/BEAM.
  """

  alias BenchElixir.Bench.Common, as: C

  # =============================================================================
  # Spawn Benchmarks
  # =============================================================================

  @doc """
  Spawn N processes that exit immediately.
  Measures raw spawn throughput without synchronization.
  """
  def spawn_n_immediate_exit(n) do
    time_ms =
      C.time_ms(fn ->
        Enum.each(1..n, fn _ ->
          spawn(fn -> :done end)
        end)
      end)

    rate = n / (time_ms / 1000.0)
    C.println_result("spawn-immediate-exit n=#{n}", rate, "proc/sec")
  end

  @doc """
  Spawn N processes, wait for all to exit using counter.
  Measures spawn + exit coordination throughput.
  """
  def spawn_n_wait_exit(n) do
    parent = self()

    time_ms =
      C.time_ms(fn ->
        Enum.each(1..n, fn _ ->
          spawn(fn -> send(parent, :done) end)
        end)

        # Wait for all processes to signal completion
        Enum.each(1..n, fn _ ->
          receive do
            :done -> :ok
          after
            60_000 -> :timeout
          end
        end)
      end)

    rate = n / (time_ms / 1000.0)
    C.println_result("spawn-wait-exit n=#{n}", rate, "proc/sec")
  end

  @doc """
  Spawn one process, wait for it to exit, repeat N times.
  Measures spawn latency (round-trip time).
  """
  def spawn_sequential(n) do
    parent = self()

    time_ms =
      C.time_ms(fn ->
        Enum.each(1..n, fn _ ->
          spawn(fn -> send(parent, :done) end)

          receive do
            :done -> :ok
          after
            1_000 -> :timeout
          end
        end)
      end)

    latency = time_ms / n
    C.println_result("spawn-sequential n=#{n}", latency, "ms/spawn")
  end

  @doc """
  Create binary tree of processes, depth D -> 2^(D+1)-1 processes.
  Each node spawns two children, waits for results, reports sum to parent.
  """
  def spawn_tree(depth) do
    parent = self()

    time_ms =
      C.time_ms(fn ->
        spawn(fn ->
          node(parent, depth)
        end)

        receive do
          {:total, total} -> total
        after
          120_000 -> :timeout
        end
      end)

    total = round(:math.pow(2, depth + 1) - 1)
    rate = total / (time_ms / 1000.0)
    C.println_result("spawn-tree depth=#{depth} procs=#{total}", rate, "proc/sec")
  end

  defp node(parent, 0) do
    send(parent, {:total, 1})
  end

  defp node(parent, d) do
    me = self()
    spawn(fn -> node(me, d - 1) end)
    spawn(fn -> node(me, d - 1) end)

    n1 =
      receive do
        {:total, n} -> n
      after
        60_000 -> 0
      end

    n2 =
      receive do
        {:total, n} -> n
      after
        60_000 -> 0
      end

    send(parent, {:total, 1 + n1 + n2})
  end

  @doc """
  Same as spawn_tree but with linked processes.
  Measures link setup overhead.
  """
  def spawn_link_tree(depth) do
    parent = self()

    time_ms =
      C.time_ms(fn ->
        spawn(fn ->
          node_linked(parent, depth)
        end)

        receive do
          {:total, total} -> total
        after
          120_000 -> :timeout
        end
      end)

    total = round(:math.pow(2, depth + 1) - 1)
    rate = total / (time_ms / 1000.0)
    C.println_result("spawn-link-tree depth=#{depth} procs=#{total}", rate, "proc/sec")
  end

  defp node_linked(parent, 0) do
    send(parent, {:total, 1})
  end

  defp node_linked(parent, d) do
    me = self()
    spawn_link(fn -> node_linked(me, d - 1) end)
    spawn_link(fn -> node_linked(me, d - 1) end)

    n1 =
      receive do
        {:total, n} -> n
      after
        60_000 -> 0
      end

    n2 =
      receive do
        {:total, n} -> n
      after
        60_000 -> 0
      end

    send(parent, {:total, 1 + n1 + n2})
  end

  @doc """
  Spawn N processes using spawn_monitor.
  Measures monitor setup overhead.
  """
  def spawn_n_monitored(n) do
    time_ms =
      C.time_ms(fn ->
        refs =
          Enum.map(1..n, fn _ ->
            {_pid, ref} = spawn_monitor(fn -> :done end)
            ref
          end)

        # Wait for all DOWN messages
        Enum.each(refs, fn ref ->
          receive do
            {:DOWN, ^ref, :process, _pid, _reason} -> :ok
          after
            60_000 -> :timeout
          end
        end)
      end)

    rate = n / (time_ms / 1000.0)
    C.println_result("spawn-monitored n=#{n}", rate, "proc/sec")
  end

  # =============================================================================
  # Runner
  # =============================================================================

  def run_all do
    C.println_header("SPAWN BENCHMARKS")

    C.println_subheader("Immediate Exit (fire-and-forget)")
    Enum.each(C.spawn_scales(), &spawn_n_immediate_exit/1)

    C.println_subheader("Wait for Exit (synchronized)")
    Enum.each(C.spawn_scales(), &spawn_n_wait_exit/1)

    C.println_subheader("Sequential (latency)")
    Enum.each(C.spawn_scales_small(), &spawn_sequential/1)

    C.println_subheader("Process Tree")
    Enum.each(C.tree_depths(), &spawn_tree/1)

    C.println_subheader("Linked Process Tree")
    Enum.each(C.tree_depths(), &spawn_link_tree/1)

    C.println_subheader("Monitored Spawns")
    Enum.each(C.spawn_scales(), &spawn_n_monitored/1)
  end
end
